File: src/data/geometry.js
import { deepCopy } from "./schema.js";
import { math } from "../utils/math.js";
import { validatePolygonLoops } from "../utils/polygon.js";

export class Geometry {
  zMin() {
    throw new Error("zMin must be implemented by subclasses");
  }

  zMax() {
    throw new Error("zMax must be implemented by subclasses");
  }

  thk() {
    throw new Error("thk must be implemented by subclasses");
  }

  copy() {
    throw new Error("copy must be implemented by subclasses");
  }

  copyWithThk() {
    throw new Error("copyWithThk must be implemented by subclasses");
  }

  copyWithXYInset() {
    throw new Error("copyWithXYInset must be implemented by subclasses");
  }

  move() {
    throw new Error("move must be implemented by subclasses");
  }

  clipTopTo() {
    throw new Error("clipTopTo must be implemented by subclasses");
  }

  clipXYToBox() {
    throw new Error("clipXYToBox must be implemented by subclasses");
  }

  flip() {
    throw new Error("flip must be implemented by subclasses");
  }

  json() {
    throw new Error("json must be implemented by subclasses");
  }

  toJSON() {
    return this.json();
  }

  z_min() {
    return this.zMin();
  }

  z_max() {
    return this.zMax();
  }

  copy_with_thk(thk) {
    return this.copyWithThk(thk);
  }

  copy_with_xy_inset(inset) {
    return this.copyWithXYInset(inset);
  }

  clip_top_to(toZ) {
    return this.clipTopTo(toZ);
  }

  clip_xy_to_box(bounds) {
    return this.clipXYToBox(bounds);
  }
}

export class BoxGeometry extends Geometry {
  constructor(bottomLeft, topRight, thk) {
    super();
    if (bottomLeft[2] !== topRight[2]) {
      throw new Error(
        "BoxGeometry bottom_left and top_right must be on the same xy plane.",
      );
    }

    this._bottomLeft = deepCopy(bottomLeft);
    this._topRight = deepCopy(topRight);
    this._thk = thk;
  }

  bottomLeft() {
    return deepCopy(this._bottomLeft);
  }

  bottom_left() {
    return this.bottomLeft();
  }

  topRight() {
    return deepCopy(this._topRight);
  }

  top_right() {
    return this.topRight();
  }

  zMin() {
    return Math.min(this._bottomLeft[2], this._topRight[2]);
  }

  zMax() {
    return this.zMin() + this._thk;
  }

  thk() {
    return this._thk;
  }

  copy() {
    return new BoxGeometry(this._bottomLeft, this._topRight, this._thk);
  }

  copyWithThk(thk) {
    return new BoxGeometry(this._bottomLeft, this._topRight, thk);
  }

  copyWithXYInset(inset) {
    const amount = finiteNumber(inset, "inset");
    const xMin = Math.min(this._bottomLeft[0], this._topRight[0]) + amount;
    const xMax = Math.max(this._bottomLeft[0], this._topRight[0]) - amount;
    const yMin = Math.min(this._bottomLeft[1], this._topRight[1]) + amount;
    const yMax = Math.max(this._bottomLeft[1], this._topRight[1]) - amount;

    if (xMin >= xMax || yMin >= yMax) {
      throw new Error("BoxGeometry XY inset collapses the footprint");
    }

    return new BoxGeometry([xMin, yMin, this.zMin()], [xMax, yMax, this.zMin()], this._thk);
  }

  move(...args) {
    const { x, y, z } = moveArgs(args);
    this._bottomLeft[0] += x;
    this._topRight[0] += x;
    this._bottomLeft[1] += y;
    this._topRight[1] += y;
    this._bottomLeft[2] += z;
    this._topRight[2] += z;
  }

  clipTopTo(toZ) {
    const zBottom = this.zMin();
    const zTop = zBottom + this._thk;

    if (math.fLe(toZ, zBottom)) return false;
    if (math.fLt(zBottom, toZ) && math.fLt(toZ, zTop)) {
      this._thk = toZ - zBottom;
    }

    return true;
  }

  clipXYToBox(bounds) {
    const crop = normalizeCropBox(bounds);
    const xMin = Math.max(
      Math.min(this._bottomLeft[0], this._topRight[0]),
      crop.xMin,
    );
    const xMax = Math.min(
      Math.max(this._bottomLeft[0], this._topRight[0]),
      crop.xMax,
    );
    const yMin = Math.max(
      Math.min(this._bottomLeft[1], this._topRight[1]),
      crop.yMin,
    );
    const yMax = Math.min(
      Math.max(this._bottomLeft[1], this._topRight[1]),
      crop.yMax,
    );

    if (math.fLe(xMax, xMin) || math.fLe(yMax, yMin)) return false;

    const z = this.zMin();
    this._bottomLeft = [xMin, yMin, z];
    this._topRight = [xMax, yMax, z];
    return true;
  }

  flip(aroundZ = 0) {
    const flippedZ = 2 * aroundZ - this.zMax();
    this._bottomLeft[2] = flippedZ;
    this._topRight[2] = flippedZ;
  }

  json() {
    return {
      type: "BoxGeometry",
      bottom_left: deepCopy(this._bottomLeft),
      top_right: deepCopy(this._topRight),
      thk: this._thk,
    };
  }
}

export class PolygonGeometry extends Geometry {
  constructor(polys, thk) {
    super();
    validatePolygonLoops(polys);
    this._polys = deepCopy(polys);
    this._thk = thk;
  }

  polygons() {
    return deepCopy(this._polys);
  }

  zMin() {
    return this._polys[0][0][2];
  }

  zMax() {
    return this.zMin() + this._thk;
  }

  thk() {
    return this._thk;
  }

  copy() {
    return new PolygonGeometry(this._polys, this._thk);
  }

  copyWithThk(thk) {
    return new PolygonGeometry(this._polys, thk);
  }

  copyWithXYInset(inset) {
    const amount = finiteNumber(inset, "inset");
    if (amount === 0) {
      return this.copy();
    }
    throw new Error("PolygonGeometry does not support non-zero XY inset");
  }

  move(...args) {
    const { x, y, z } = moveArgs(args);
    this._polys.forEach((poly) => {
      poly.forEach((node) => {
        node[0] += x;
        node[1] += y;
        node[2] += z;
      });
    });
  }

  clipTopTo(toZ) {
    const zBottom = this.zMin();
    const zTop = zBottom + this._thk;

    if (math.fLe(toZ, zBottom)) return false;
    if (math.fLt(zBottom, toZ) && math.fLt(toZ, zTop)) {
      this._thk = toZ - zBottom;
    }

    return true;
  }

  clipXYToBox(bounds) {
    const crop = normalizeCropBox(bounds);
    const clipped = this._polys
      .map((poly) => clipLoopToBox(poly, crop))
      .filter((poly) => poly.length >= 3);

    if (clipped.length === 0) return false;

    validatePolygonLoops(clipped);
    this._polys = clipped;
    return true;
  }

  flip(aroundZ = 0) {
    const flippedZ = 2 * aroundZ - this.zMax();
    this._polys.forEach((poly) => {
      poly.forEach((node) => {
        node[2] = flippedZ;
      });
    });
  }

  json() {
    return {
      type: "PolygonGeometry",
      polys: deepCopy(this._polys),
      thk: this._thk,
    };
  }
}

export class CylinderGeometry extends Geometry {
  constructor(center, bottomRadius, thk) {
    super();
    this._center = deepCopy(center);
    this._bottomRadius = bottomRadius;
    this._thk = thk;
  }

  center() {
    return deepCopy(this._center);
  }

  bottomRadius() {
    return this._bottomRadius;
  }

  bottom_radius() {
    return this.bottomRadius();
  }

  zMin() {
    return this._center[2];
  }

  zMax() {
    return this.zMin() + this._thk;
  }

  thk() {
    return this._thk;
  }

  copy() {
    return new CylinderGeometry(this._center, this._bottomRadius, this._thk);
  }

  copyWithThk(thk) {
    return new CylinderGeometry(this._center, this._bottomRadius, thk);
  }

  copyWithXYInset(inset) {
    const radius = this._bottomRadius - finiteNumber(inset, "inset");
    if (radius <= 0) {
      throw new Error("CylinderGeometry XY inset collapses the footprint");
    }
    return new CylinderGeometry(this._center, radius, this._thk);
  }

  move(...args) {
    const { x, y, z } = moveArgs(args);
    this._center[0] += x;
    this._center[1] += y;
    this._center[2] += z;
  }

  clipTopTo(toZ) {
    const zBottom = this.zMin();
    const zTop = zBottom + this._thk;

    if (math.fLe(toZ, zBottom)) return false;
    if (math.fLt(zBottom, toZ) && math.fLt(toZ, zTop)) {
      this._thk = toZ - zBottom;
    }

    return true;
  }

  clipXYToBox(bounds) {
    return clipCircularFootprintToBox({
      bounds,
      center: this._center,
      radius: this._bottomRadius,
      typeName: "CylinderGeometry",
    });
  }

  flip(aroundZ = 0) {
    this._center[2] = 2 * aroundZ - this.zMax();
  }

  json() {
    return {
      type: "CylinderGeometry",
      center: deepCopy(this._center),
      bottom_radius: this._bottomRadius,
      thk: this._thk,
    };
  }
}

export class ConeGeometry extends Geometry {
  constructor(center, bottomRadius, topRadius, thk) {
    super();
    this._center = deepCopy(center);
    this._bottomRadius = bottomRadius;
    this._topRadius = topRadius;
    this._thk = thk;
  }

  center() {
    return deepCopy(this._center);
  }

  bottomRadius() {
    return this._bottomRadius;
  }

  bottom_radius() {
    return this.bottomRadius();
  }

  topRadius() {
    return this._topRadius;
  }

  top_radius() {
    return this.topRadius();
  }

  zMin() {
    return this._center[2];
  }

  zMax() {
    return this.zMin() + this._thk;
  }

  thk() {
    return this._thk;
  }

  copy() {
    return new ConeGeometry(
      this._center,
      this._bottomRadius,
      this._topRadius,
      this._thk,
    );
  }

  copyWithThk(thk) {
    return new ConeGeometry(
      this._center,
      this._bottomRadius,
      this._topRadius,
      thk,
    );
  }

  copyWithXYInset(inset) {
    const amount = finiteNumber(inset, "inset");
    const bottomRadius = this._bottomRadius - amount;
    const topRadius = this._topRadius - amount;
    if (bottomRadius <= 0 || topRadius <= 0) {
      throw new Error("ConeGeometry XY inset collapses the footprint");
    }
    return new ConeGeometry(this._center, bottomRadius, topRadius, this._thk);
  }

  move(...args) {
    const { x, y, z } = moveArgs(args);
    this._center[0] += x;
    this._center[1] += y;
    this._center[2] += z;
  }

  clipTopTo(toZ) {
    const zBottom = this.zMin();
    const zTop = zBottom + this._thk;

    if (math.fLe(toZ, zBottom)) return false;
    if (math.fLt(zBottom, toZ) && math.fLt(toZ, zTop)) {
      this._thk = toZ - zBottom;
    }

    return true;
  }

  clipXYToBox(bounds) {
    return clipCircularFootprintToBox({
      bounds,
      center: this._center,
      radius: Math.max(this._bottomRadius, this._topRadius),
      typeName: "ConeGeometry",
    });
  }

  flip(aroundZ = 0) {
    this._center[2] = 2 * aroundZ - this.zMax();
    [this._bottomRadius, this._topRadius] = [
      this._topRadius,
      this._bottomRadius,
    ];
  }

  json() {
    return {
      type: "ConeGeometry",
      center: deepCopy(this._center),
      bottom_radius: this._bottomRadius,
      top_radius: this._topRadius,
      thk: this._thk,
    };
  }
}

export function moveArgs(args) {
  if (args.length === 0) {
    return { x: 0, y: 0, z: 0 };
  }
  if (
    args.length === 1 &&
    args[0] !== null &&
    typeof args[0] === "object" &&
    !Array.isArray(args[0])
  ) {
    const { x = 0, y = 0, z = 0 } = args[0];
    return { x, y, z };
  }
  const [x = 0, y = 0, z = 0] = args;
  return { x, y, z };
}

function finiteNumber(value, label) {
  const number = Number(value);
  if (!Number.isFinite(number)) {
    throw new Error(`${label} must be a finite number`);
  }
  return number;
}

function normalizeCropBox(bounds) {
  const xMin = finiteNumber(bounds?.xMin, "bounds.xMin");
  const xMax = finiteNumber(bounds?.xMax, "bounds.xMax");
  const yMin = finiteNumber(bounds?.yMin, "bounds.yMin");
  const yMax = finiteNumber(bounds?.yMax, "bounds.yMax");
  if (math.fLe(xMax, xMin) || math.fLe(yMax, yMin)) {
    throw new Error("clipXYToBox requires a non-empty XY box");
  }
  return { xMin, xMax, yMin, yMax };
}

function clipCircularFootprintToBox({ bounds, center, radius, typeName }) {
  const crop = normalizeCropBox(bounds);
  const r = finiteNumber(radius, "radius");
  const x = center[0];
  const y = center[1];

  if (
    math.fGe(x - r, crop.xMin) &&
    math.fLe(x + r, crop.xMax) &&
    math.fGe(y - r, crop.yMin) &&
    math.fLe(y + r, crop.yMax)
  ) {
    return true;
  }

  const closestX = Math.min(Math.max(x, crop.xMin), crop.xMax);
  const closestY = Math.min(Math.max(y, crop.yMin), crop.yMax);
  const dx = x - closestX;
  const dy = y - closestY;
  if (math.fGe(dx * dx + dy * dy, r * r)) {
    return false;
  }

  throw new Error(`${typeName} does not support partial XY saw clipping`);
}

function clipLoopToBox(loop, bounds) {
  const z = loop[0][2];
  let result = loop.map((point) => [...point]);
  result = clipLoopAgainstBoundary(result, bounds, insideLeft, intersectLeft);
  result = clipLoopAgainstBoundary(result, bounds, insideRight, intersectRight);
  result = clipLoopAgainstBoundary(result, bounds, insideBottom, intersectBottom);
  result = clipLoopAgainstBoundary(result, bounds, insideTop, intersectTop);
  return cleanLoop(result.map(([x, y]) => [math.fZero(x), math.fZero(y), z]));
}

function clipLoopAgainstBoundary(loop, bounds, inside, intersect) {
  if (loop.length === 0) return [];
  const result = [];
  let previous = loop[loop.length - 1];
  let previousInside = inside(previous, bounds);

  for (const current of loop) {
    const currentInside = inside(current, bounds);
    if (currentInside) {
      if (!previousInside) {
        result.push(intersect(previous, current, bounds));
      }
      result.push(current);
    } else if (previousInside) {
      result.push(intersect(previous, current, bounds));
    }
    previous = current;
    previousInside = currentInside;
  }

  return cleanAdjacentDuplicates(result);
}

function insideLeft(point, bounds) {
  return math.fGe(point[0], bounds.xMin);
}

function insideRight(point, bounds) {
  return math.fLe(point[0], bounds.xMax);
}

function insideBottom(point, bounds) {
  return math.fGe(point[1], bounds.yMin);
}

function insideTop(point, bounds) {
  return math.fLe(point[1], bounds.yMax);
}

function intersectLeft(start, end, bounds) {
  return intersectAtX(start, end, bounds.xMin);
}

function intersectRight(start, end, bounds) {
  return intersectAtX(start, end, bounds.xMax);
}

function intersectBottom(start, end, bounds) {
  return intersectAtY(start, end, bounds.yMin);
}

function intersectTop(start, end, bounds) {
  return intersectAtY(start, end, bounds.yMax);
}

function intersectAtX(start, end, x) {
  const dx = end[0] - start[0];
  if (math.fEq(dx, 0)) return [x, start[1], start[2]];
  const t = (x - start[0]) / dx;
  return [x, start[1] + t * (end[1] - start[1]), start[2]];
}

function intersectAtY(start, end, y) {
  const dy = end[1] - start[1];
  if (math.fEq(dy, 0)) return [start[0], y, start[2]];
  const t = (y - start[1]) / dy;
  return [start[0] + t * (end[0] - start[0]), y, start[2]];
}

function cleanLoop(loop) {
  const cleaned = cleanAdjacentDuplicates(loop);
  if (
    cleaned.length > 1 &&
    samePoint2(cleaned[0], cleaned[cleaned.length - 1])
  ) {
    cleaned.pop();
  }
  if (cleaned.length < 3 || math.fEq(signedArea2(cleaned), 0)) {
    return [];
  }
  return cleaned;
}

function cleanAdjacentDuplicates(loop) {
  const result = [];
  for (const point of loop) {
    if (
      result.length === 0 ||
      !samePoint2(result[result.length - 1], point)
    ) {
      result.push(point);
    }
  }
  return result;
}

function signedArea2(loop) {
  let area = 0;
  for (let index = 0; index < loop.length; index += 1) {
    const point = loop[index];
    const nextPoint = loop[(index + 1) % loop.length];
    area += point[0] * nextPoint[1] - nextPoint[0] * point[1];
  }
  return area / 2;
}

function samePoint2(left, right) {
  return math.fEq(left[0], right[0]) && math.fEq(left[1], right[1]);
}

File: packages/kernel-py/src/process_flow_kernel/domain/geometry.py
from __future__ import annotations

from ..serialization.schema import deep_copy
from ..utils.math_utils import math
from ..utils.polygon import clip_polygon_loops_to_box, validate_polygon_loops


class Geometry:
    def z_min(self):
        raise NotImplementedError

    def z_max(self):
        raise NotImplementedError

    def thk(self):
        raise NotImplementedError

    def copy(self):
        raise NotImplementedError

    def copy_with_thk(self, thk):
        raise NotImplementedError

    def copy_with_xy_inset(self, inset):
        raise NotImplementedError

    def move(self, x=0, y=0, z=0):
        raise NotImplementedError

    def resize_xy_by(self, delta_x, delta_y):
        raise NotImplementedError

    def clip_top_to(self, to_z):
        raise NotImplementedError

    def clip_xy_to_box(self, bounds):
        """Return the retained geometry, or None when the crop removes it."""
        raise NotImplementedError

    def flip(self, around_z=0):
        raise NotImplementedError

    def json(self):
        raise NotImplementedError


class BoxGeometry(Geometry):
    def __init__(self, bottom_left, top_right, thk):
        if bottom_left[2] != top_right[2]:
            raise ValueError("BoxGeometry bottom_left and top_right must be on the same xy plane.")
        self._bottom_left = deep_copy(bottom_left)
        self._top_right = deep_copy(top_right)
        self._thk = thk

    def bottom_left(self):
        return deep_copy(self._bottom_left)

    def top_right(self):
        return deep_copy(self._top_right)

    def z_min(self):
        return min(self._bottom_left[2], self._top_right[2])

    def z_max(self):
        return self.z_min() + self._thk

    def thk(self):
        return self._thk

    def copy(self):
        return BoxGeometry(self._bottom_left, self._top_right, self._thk)

    def copy_with_thk(self, thk):
        return BoxGeometry(self._bottom_left, self._top_right, thk)

    def copy_with_xy_inset(self, inset):
        amount = _finite_number(inset, "inset")
        x_min = min(self._bottom_left[0], self._top_right[0]) + amount
        x_max = max(self._bottom_left[0], self._top_right[0]) - amount
        y_min = min(self._bottom_left[1], self._top_right[1]) + amount
        y_max = max(self._bottom_left[1], self._top_right[1]) - amount
        if x_min >= x_max or y_min >= y_max:
            raise ValueError("BoxGeometry XY inset collapses the footprint")
        return BoxGeometry([x_min, y_min, self.z_min()], [x_max, y_max, self.z_min()], self._thk)

    def move(self, x=0, y=0, z=0):
        self._bottom_left[0] += x
        self._top_right[0] += x
        self._bottom_left[1] += y
        self._top_right[1] += y
        self._bottom_left[2] += z
        self._top_right[2] += z

    def resize_xy_by(self, delta_x, delta_y):
        resized_x = self._top_right[0] + _finite_number(delta_x, "deltaX")
        resized_y = self._top_right[1] + _finite_number(delta_y, "deltaY")
        if math.f_le(resized_x, self._bottom_left[0]) or math.f_le(
            resized_y, self._bottom_left[1]
        ):
            raise ValueError("BoxGeometry XY resize collapses the footprint")
        self._top_right[0] = resized_x
        self._top_right[1] = resized_y

    def clip_top_to(self, to_z):
        z_bottom = self.z_min()
        z_top = z_bottom + self._thk
        if math.f_le(to_z, z_bottom):
            return False
        if math.f_lt(z_bottom, to_z) and math.f_lt(to_z, z_top):
            self._thk = to_z - z_bottom
        return True

    def clip_xy_to_box(self, bounds):
        crop = _normalize_crop_box(bounds)
        x_min = max(min(self._bottom_left[0], self._top_right[0]), crop["xMin"])
        x_max = min(max(self._bottom_left[0], self._top_right[0]), crop["xMax"])
        y_min = max(min(self._bottom_left[1], self._top_right[1]), crop["yMin"])
        y_max = min(max(self._bottom_left[1], self._top_right[1]), crop["yMax"])
        if math.f_le(x_max, x_min) or math.f_le(y_max, y_min):
            return None
        z = self.z_min()
        self._bottom_left = [x_min, y_min, z]
        self._top_right = [x_max, y_max, z]
        return self

    def flip(self, around_z=0):
        flipped_z = 2 * around_z - self.z_max()
        self._bottom_left[2] = flipped_z
        self._top_right[2] = flipped_z

    def json(self):
        return {
            "type": "BoxGeometry",
            "bottom_left": deep_copy(self._bottom_left),
            "top_right": deep_copy(self._top_right),
            "thk": self._thk,
        }


class PolygonGeometry(Geometry):
    def __init__(self, polys, thk):
        validate_polygon_loops(polys)
        self._polys = deep_copy(polys)
        self._thk = thk

    def polygons(self):
        return deep_copy(self._polys)

    def z_min(self):
        return self._polys[0][0][2]

    def z_max(self):
        return self.z_min() + self._thk

    def thk(self):
        return self._thk

    def copy(self):
        return PolygonGeometry(self._polys, self._thk)

    def copy_with_thk(self, thk):
        return PolygonGeometry(self._polys, thk)

    def copy_with_xy_inset(self, inset):
        amount = _finite_number(inset, "inset")
        if amount == 0:
            return self.copy()
        raise ValueError("PolygonGeometry does not support non-zero XY inset")

    def move(self, x=0, y=0, z=0):
        for poly in self._polys:
            for node in poly:
                node[0] += x
                node[1] += y
                node[2] += z

    def resize_xy_by(self, delta_x, delta_y):
        _ = delta_x, delta_y
        raise ValueError("PnP XY resize supports only BoxGeometry")

    def clip_top_to(self, to_z):
        z_bottom = self.z_min()
        z_top = z_bottom + self._thk
        if math.f_le(to_z, z_bottom):
            return False
        if math.f_lt(z_bottom, to_z) and math.f_lt(to_z, z_top):
            self._thk = to_z - z_bottom
        return True

    def clip_xy_to_box(self, bounds):
        crop = _normalize_crop_box(bounds)
        if all(
            math.f_ge(point[0], crop["xMin"])
            and math.f_le(point[0], crop["xMax"])
            and math.f_ge(point[1], crop["yMin"])
            and math.f_le(point[1], crop["yMax"])
            for poly in self._polys
            for point in poly
        ):
            return self
        clipped = clip_polygon_loops_to_box(self._polys, crop)
        if len(clipped) == 0:
            return None
        self._polys = clipped
        return self

    def flip(self, around_z=0):
        flipped_z = 2 * around_z - self.z_max()
        for poly in self._polys:
            for node in poly:
                node[2] = flipped_z

    def json(self):
        return {
            "type": "PolygonGeometry",
            "polys": deep_copy(self._polys),
            "thk": self._thk,
        }


class CylinderGeometry(Geometry):
    def __init__(self, center, bottom_radius, thk):
        self._center = deep_copy(center)
        self._bottom_radius = bottom_radius
        self._thk = thk

    def center(self):
        return deep_copy(self._center)

    def bottom_radius(self):
        return self._bottom_radius

    def z_min(self):
        return self._center[2]

    def z_max(self):
        return self.z_min() + self._thk

    def thk(self):
        return self._thk

    def copy(self):
        return CylinderGeometry(self._center, self._bottom_radius, self._thk)

    def copy_with_thk(self, thk):
        return CylinderGeometry(self._center, self._bottom_radius, thk)

    def copy_with_xy_inset(self, inset):
        radius = self._bottom_radius - _finite_number(inset, "inset")
        if radius <= 0:
            raise ValueError("CylinderGeometry XY inset collapses the footprint")
        return CylinderGeometry(self._center, radius, self._thk)

    def move(self, x=0, y=0, z=0):
        self._center[0] += x
        self._center[1] += y
        self._center[2] += z

    def resize_xy_by(self, delta_x, delta_y):
        _ = delta_x, delta_y
        raise ValueError("PnP XY resize supports only BoxGeometry")

    def clip_top_to(self, to_z):
        z_bottom = self.z_min()
        z_top = z_bottom + self._thk
        if math.f_le(to_z, z_bottom):
            return False
        if math.f_lt(z_bottom, to_z) and math.f_lt(to_z, z_top):
            self._thk = to_z - z_bottom
        return True

    def clip_xy_to_box(self, bounds):
        crop = _normalize_crop_box(bounds)
        relation = _circular_footprint_box_relation(
            crop=crop, center=self._center, radius=self._bottom_radius
        )
        if relation == "circleInsideBox":
            return self
        if relation == "disjoint":
            return None
        if _box_is_inside_circle(crop=crop, center=self._center, radius=self._bottom_radius):
            return BoxGeometry(
                [crop["xMin"], crop["yMin"], self.z_min()],
                [crop["xMax"], crop["yMax"], self.z_min()],
                self._thk,
            )
        raise ValueError("CylinderGeometry does not support partial XY saw clipping")

    def flip(self, around_z=0):
        self._center[2] = 2 * around_z - self.z_max()

    def json(self):
        return {
            "type": "CylinderGeometry",
            "center": deep_copy(self._center),
            "bottom_radius": self._bottom_radius,
            "thk": self._thk,
        }


class ConeGeometry(Geometry):
    def __init__(self, center, bottom_radius, top_radius, thk):
        self._center = deep_copy(center)
        self._bottom_radius = bottom_radius
        self._top_radius = top_radius
        self._thk = thk

    def center(self):
        return deep_copy(self._center)

    def bottom_radius(self):
        return self._bottom_radius

    def top_radius(self):
        return self._top_radius

    def z_min(self):
        return self._center[2]

    def z_max(self):
        return self.z_min() + self._thk

    def thk(self):
        return self._thk

    def copy(self):
        return ConeGeometry(self._center, self._bottom_radius, self._top_radius, self._thk)

    def copy_with_thk(self, thk):
        return ConeGeometry(self._center, self._bottom_radius, self._top_radius, thk)

    def copy_with_xy_inset(self, inset):
        amount = _finite_number(inset, "inset")
        bottom_radius = self._bottom_radius - amount
        top_radius = self._top_radius - amount
        if bottom_radius <= 0 or top_radius <= 0:
            raise ValueError("ConeGeometry XY inset collapses the footprint")
        return ConeGeometry(self._center, bottom_radius, top_radius, self._thk)

    def move(self, x=0, y=0, z=0):
        self._center[0] += x
        self._center[1] += y
        self._center[2] += z

    def resize_xy_by(self, delta_x, delta_y):
        _ = delta_x, delta_y
        raise ValueError("PnP XY resize supports only BoxGeometry")

    def clip_top_to(self, to_z):
        z_bottom = self.z_min()
        z_top = z_bottom + self._thk
        if math.f_le(to_z, z_bottom):
            return False
        if math.f_lt(z_bottom, to_z) and math.f_lt(to_z, z_top):
            self._thk = to_z - z_bottom
        return True

    def clip_xy_to_box(self, bounds):
        relation = _circular_footprint_box_relation(
            crop=_normalize_crop_box(bounds),
            center=self._center,
            radius=max(self._bottom_radius, self._top_radius),
        )
        if relation == "circleInsideBox":
            return self
        if relation == "disjoint":
            return None
        raise ValueError("ConeGeometry does not support partial XY saw clipping")

    def flip(self, around_z=0):
        self._center[2] = 2 * around_z - self.z_max()
        self._bottom_radius, self._top_radius = self._top_radius, self._bottom_radius

    def json(self):
        return {
            "type": "ConeGeometry",
            "center": deep_copy(self._center),
            "bottom_radius": self._bottom_radius,
            "top_radius": self._top_radius,
            "thk": self._thk,
        }


def _finite_number(value, label):
    try:
        number = float(value)
    except (TypeError, ValueError):
        raise ValueError(f"{label} must be a finite number") from None
    if number in (float("inf"), float("-inf")) or number != number:
        raise ValueError(f"{label} must be a finite number")
    return number


def _normalize_crop_box(bounds):
    x_min = _finite_number(bounds.get("xMin"), "bounds.xMin")
    x_max = _finite_number(bounds.get("xMax"), "bounds.xMax")
    y_min = _finite_number(bounds.get("yMin"), "bounds.yMin")
    y_max = _finite_number(bounds.get("yMax"), "bounds.yMax")
    if math.f_le(x_max, x_min) or math.f_le(y_max, y_min):
        raise ValueError("clipXYToBox requires a non-empty XY box")
    return {"xMin": x_min, "xMax": x_max, "yMin": y_min, "yMax": y_max}


def _circular_footprint_box_relation(*, crop, center, radius):
    r = _finite_number(radius, "radius")
    x = center[0]
    y = center[1]
    if (
        math.f_ge(x - r, crop["xMin"])
        and math.f_le(x + r, crop["xMax"])
        and math.f_ge(y - r, crop["yMin"])
        and math.f_le(y + r, crop["yMax"])
    ):
        return "circleInsideBox"

    closest_x = min(max(x, crop["xMin"]), crop["xMax"])
    closest_y = min(max(y, crop["yMin"]), crop["yMax"])
    dx = x - closest_x
    dy = y - closest_y
    if math.f_ge(dx * dx + dy * dy, r * r):
        return "disjoint"

    return "partial"


def _box_is_inside_circle(*, crop, center, radius):
    r_squared = _finite_number(radius, "radius") ** 2
    x = center[0]
    y = center[1]
    return all(
        math.f_le((corner_x - x) ** 2 + (corner_y - y) ** 2, r_squared)
        for corner_x, corner_y in (
            (crop["xMin"], crop["yMin"]),
            (crop["xMin"], crop["yMax"]),
            (crop["xMax"], crop["yMin"]),
            (crop["xMax"], crop["yMax"]),
        )
    )
